fuzzer/src/main/scala/firrtl/ExprGen.scala

```diff
@@ -0,0 +1,728 @@
+        }
+      }
+    }
+
+    def boolSIntGen[S: ExprState, G[_]: GenMonad]: Option[StateGen[S, G, DoPrim]] = None
+    def sintGen[S: ExprState, G[_]: GenMonad]: Option[Width => StateGen[S, G, DoPrim]] = None
+  }
+
+  object AsUIntDoPrimGen extends DoPrimGen(PrimOps.AsUInt) {
+    def boolUIntGen[S: ExprState, G[_]: GenMonad]: Option[StateGen[S, G, DoPrim]] = uintGen.map(_(1))
+    def uintGen[S: ExprState, G[_]: GenMonad]: Option[Width => StateGen[S, G, DoPrim]] = {
+      Some { width =>
+        val intWidth = IntWidth(width)
+        for {
+          tpe <- StateGen.liftG(GenMonad[G].oneOf(UIntType(_), SIntType(_)))
+          arg <- ExprState[S].exprGen(tpe(intWidth))
+        } yield {
+          DoPrim(primOp, Seq(arg), Seq(), UIntType(intWidth))
+        }
+      }
+    }
+
+    def boolSIntGen[S: ExprState, G[_]: GenMonad]: Option[StateGen[S, G, DoPrim]] = None
+    def sintGen[S: ExprState, G[_]: GenMonad]: Option[Width => StateGen[S, G, DoPrim]] = None
+  }
+
+  object AsSIntDoPrimGen extends DoPrimGen(PrimOps.AsSInt) {
+    def boolUIntGen[S: ExprState, G[_]: GenMonad]: Option[StateGen[S, G, DoPrim]] = None
+    def uintGen[S: ExprState, G[_]: GenMonad]: Option[Width => StateGen[S, G, DoPrim]] = None
+
+    def boolSIntGen[S: ExprState, G[_]: GenMonad]: Option[StateGen[S, G, DoPrim]] = sintGen.map(_(1))
+    def sintGen[S: ExprState, G[_]: GenMonad]: Option[Width => StateGen[S, G, DoPrim]] = {
+      Some { width =>
+        val intWidth = IntWidth(width)
+        for {
+          tpe <- StateGen.liftG(GenMonad[G].oneOf(UIntType(_), SIntType(_)))
+          arg <- ExprState[S].exprGen(tpe(intWidth))
+        } yield {
+          DoPrim(primOp, Seq(arg), Seq(), SIntType(intWidth))
+        }
+      }
+    }
+  }
+
+  object MuxGen extends ExprGen[Mux] {
+    def name = "mux"
+    private def imp[S: ExprState, G[_]: GenMonad](isUInt: Boolean)(width: BigInt): StateGen[S, G, Mux] = {
+      val tpe = if (isUInt) UIntType(_) else SIntType(_)
+      // spec states that types must be equivalent, but in practice we allow differing widths
+      for {
+        cond <- ExprState[S].exprGen(Utils.BoolType)
+        flip <- StateGen.liftG(GenMonad.bool)
+        expr1 <- ExprState[S].exprGen(tpe(IntWidth(width)))
+        width2 <- StateGen.liftG(genWidthMax(width.toInt))
+        expr2 <- ExprState[S].exprGen(tpe(width2))
+      } yield {
+        Mux(cond, expr1, expr2, tpe(IntWidth(width)))
+      }
+    }
+    def boolUIntGen[S: ExprState, G[_]: GenMonad]: Option[StateGen[S, G, Mux]] = uintGen.map(_(1))
+    def uintGen[S: ExprState, G[_]: GenMonad]: Option[Width => StateGen[S, G, Mux]] = {
+      Some { imp(isUInt = true) }
+    }
+
+    def boolSIntGen[S: ExprState, G[_]: GenMonad]: Option[StateGen[S, G, Mux]] = sintGen.map(_(1))
+    def sintGen[S: ExprState, G[_]: GenMonad]: Option[Width => StateGen[S, G, Mux]] = {
+      Some { imp(isUInt = false) }
+    }
+  }
```

Comment:
One thing that has never really made sense to me is why there are function definitions for each `boolUInt`, `uint`, `boolSInt`, and `sint`. How is this going to work when we add aggregates? 

I highlighted this particular block fo code because the private `imp` taking `isUInt` as a `Boolean` argument kind of highlights the point. What will this code look like with Bundles and Vecs?

Reply:
There are different definitions for each because some primops like `lt` `ge`, etc. can only generate `boolUInt` and some cannot generate single bit ground types without generating zero-width wires. For aggregates I don't think there will be these edge-cases so we can just add two more methods for bundles and vecs.

In this case the code for Bundles and Vecs would look like
```
private def imp[S: ExprState, G[_]: GenMonad](tpe: Type): StateGen[S, G, Mux] = {
  for {
    cond <- ExprState[S].exprGen(Utils.BoolType)
    expr1 <- ExprState[S].exprGen(tpe)
    expr2 <- ExprState[S].exprGen(tpe)
  } yield {
    Mux(cond, expr1, expr2, tpe)
  }
}
```
This could be used for all the types, but since we allow uint/sint types with differing widths the generator is a bit different here.

Reply:
I see, alright I think this can work.